sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,473 @@
+type Request struct {
+	// pid is the identifier of the process which triggered the request.
+	pid uint32
+	// label is the apparmor label on the process which triggered the request.
+	label string
+	// subjectUID is the UID of the subject which triggered the request.
+	subjectUID uint32
+
+	// path is the path of the file, as seen by the process triggering the request.
+	path string
+	// class is the mediation class corresponding to this request.
+	class notify.MediationClass
+	// permission is the opaque permission that is being requested.
+	permission interface{}
+	// replyChan is a channel for writing the response.
+	replyChan chan interface{}
+	// replied indicates whether a reply has already been sent for this request.
+	replied uint32
+}
+
+func newRequest(msg *notify.MsgNotificationFile) (*Request, error) {
+	var perm interface{}
+	switch msg.Class {
+	case notify.AA_CLASS_FILE:
+		_, missingPerms, err := msg.DecodeFilePermissions()
+		if err != nil {
+			return nil, err
+		}
+		perm = missingPerms
+	default:
+		return nil, fmt.Errorf("unsupported mediation class: %v", msg.Class)
+	}
+	return &Request{
+		pid:        msg.Pid,
+		label:      msg.Label,
+		subjectUID: msg.SUID,
+
+		path:       msg.Name,
+		class:      msg.Class,
+		permission: perm,
+
+		replyChan: make(chan interface{}, 1),
+	}, nil
+}
+
+// PID returns the PID of the process which triggered the request.
+func (r *Request) PID() uint32 {
+	return r.pid
+}
+
+// Label returns the apparmor label on the process which triggered the request.
+func (r *Request) Label() string {
+	return r.label
+}
+
+// SubjectUID returns the UID of the subject which triggered the request.
+func (r *Request) SubjectUID() uint32 {
+	return r.subjectUID
+}
+
+// Path is the path of the file, as seen by the process which triggered the request.
+func (r *Request) Path() string {
+	return r.path
+}
+
+// Class is the mediation class corresponding to this request.
+func (r *Request) Class() notify.MediationClass {
+	return r.class
+}
+
+// Permission returns the opaque permission that is being requested.
+func (r *Request) Permission() interface{} {
+	return r.permission
+}
+
+// Reply sends the given response back to the kernel.
+func (r *Request) Reply(response interface{}) error {
+	if !atomic.CompareAndSwapUint32(&r.replied, 0, 1) {
+		return ErrAlreadyReplied
+	}
+	var ok bool
+	switch r.Class() {
+	case notify.AA_CLASS_FILE:
+		_, ok = response.(bool)
+	default:
+		// should not occur, since the request was created in this package
+		return fmt.Errorf("internal error: unsupported mediation class: %v", r.Class())
+	}
+	if !ok {
+		expectedType := expectedResponseTypeForClass(r.Class())
+		return fmt.Errorf("invalid reply: response must be of type %s", expectedType)
+	}
+	r.replyChan <- response
+	return nil
+}
+
+func expectedResponseTypeForClass(class notify.MediationClass) string {
+	switch class {
+	case notify.AA_CLASS_FILE:
+		return "bool"
+	default:
+		return "unsupported"
```

Comment:
given how the helper is used this could be a panic, "unsupported" reads a bit strangely because it is not a type, it's about class being unsupported,  we you still want return a string either "" or "?" might work better

Reply:
Good point! I'll use `???` so it's more clear that it's a string one can grep for.
There's currently no way for the this to occur, since the (only) caller returns immediately before if the class is not supported. I'll add a comment noting as much.